### ISSUE TITLE
Guidance loss 

### ISSUE THREAD

**Issue by KumapowerLIU** [Zhaoyi-Yan/Shift-Net_pytorch#16]
gt_latent = self.ng_innerCos_list[0].get_target()  # then get gt_latent(should be variable require_grad=False)
self.ng_innerCos_list[0].set_target(gt_latent)

I can  not find these codes in   set_gt_latent(self) function, so confused

**Comment by Zhaoyi-Yan** [comment]
Hi，can you tell me which line your snippet is?

**Comment by KumapowerLIU** [comment]
谢谢 麻烦了  验证后麻烦回复我一下= =

**Comment by tchaton** [comment]
@KumapowerLIU  @Zhaoyi-Yan  Please english :)

**Comment by Zhaoyi-Yan** [comment]
@KumapowerLIU Could you try the version of master? The shift operation has been validated. Any you can enjoy the fast speed of optimized shift. Maybe you can try to confirm whether `InnerCos` is in line your expectation in the master.

**Comment by tchaton** [comment]
![image](https://user-images.githubusercontent.com/12861981/50076094-a12aa100-01d8-11e9-8fc9-20a760a50e04.png)

Here is a training loss. We can certainly do something to help preventing GAN collapse.

Best,
T.C

**Comment by Zhaoyi-Yan** [comment]
`vallian gan` usually suffers from model collapse. I haven't tried other gans.
Does `relativelistic gan` or `spectral norm` helps avoid collapse?

**Comment by tchaton** [comment]
I didn t really tried yet.

**Comment by tchaton** [comment]
Hello Yann,

I am trying a new l1 loss. It is weighted sum between l1 loss inside ans outsider the mask. The better the outside will be. The more it will focus on the inside.

I also want to try inverse signed distance function applied over the mask.

 Best, 
T.C

**Comment by tchaton** [comment]
I have noticed, that it was creating a constant collapse, even when I was adding a differ starter.

Because the generator mess up outside the mask, the discriminator starts to give back shitty gradients and the Gan collapse.

Best,
T.C

**Comment by Zhaoyi-Yan** [comment]
`square mask` training makes the testing only suitable for the same square masking. To avoid this, we can probably set the square mask randomly in training. In this way, a local D is expected to distinguish the mask region. Something like the settings in `global and local` and `generative inpainting`.

**Comment by Zhaoyi-Yan** [comment]
@KumapowerLIU   I am sorry for the delay. You are totally right. It is a bug, that I should not put this line `self.loss = self.criterion(self.former_in_mask * self.strength, Variable(self.target))` at the end of `InnerCos forward.` In a very early version, https://github.com/Zhaoyi-Yan/Shift-Net_pytorch/blob/ac4853d61abf601cfe69ff0b6eb03bd8888d1aa8/models/InnerCos.py#L47-L54

`self.loss` is put before the setting of `self.target`. In this case, the guidance loss truly makes sense. Thank you for your reporting.

**Comment by Zhaoyi-Yan** [comment]
#53  Fixed, please check, thank ypu!
